### PR TITLE
Immediate Feedback fixes #145

### LINKED ISSUE #145: Indicate that branches/PR's are being used in the repository

### PR DESCRIPTION
**Pull request by LiamClark**
This pr provides immediate feedback to the student about their pull-request usage for the current assignment.

### CODE REVIEW

File: src/test/java/nl/tudelft/ewi/devhub/server/database/controllers/WarningTest.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+        @Mock
+        private Delivery delivery;
+
+
+        private WarningResource warningResource;
+
+        @Before
+        public void setup() {
+            warningResource = new WarningResource(group, null, pullRequest, deliveries, null, repository);
```
Comment (JWGmeligMeyling) [inline src/test/java/nl/tudelft/ewi/devhub/server/database/controllers/WarningTest.java]: I recently found out the [`@InjectMocks`](http://docs.mockito.googlecode.com/hg/1.9.5/org/mockito/InjectMocks.html) annotation which is perfect for this case.

File: src/main/java/nl/tudelft/ewi/devhub/server/web/resources/WarningResource.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        Stream<PullRequest> pullRequestStream = pullRequests.findAllPullRequests(group).stream();
+
+        if(lastDelivery != null) {
+            Date lastDeliveryDate = lastDelivery.getCreated();
+            pullRequestStream = pullRequestStream.filter((pr) -> lastDeliveryDate.before(mapDatePullRequest(pr)));
+        }
+        return findStatus(pullRequestStream.collect(Collectors.toList()));
+    }
+
+    public static PullRequestWarningResponse findStatus(List<PullRequest> pullRequests){
```
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/web/resources/WarningResource.java]: This helper method takes **recent** pull requests as input, the name pullRequests can be misunderstood.

File: src/main/java/nl/tudelft/ewi/devhub/server/web/resources/WarningResource.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    @Produces(MediaType.APPLICATION_JSON)
+    public PullRequestWarningResponse getPullRequestUsageStatus(){
+        Delivery lastDelivery = deliveries.getLastDelivery(group);
+        Stream<PullRequest> pullRequestStream = pullRequests.findAllPullRequests(group).stream();
+
+        if(lastDelivery != null) {
+            Date lastDeliveryDate = lastDelivery.getCreated();
+            pullRequestStream = pullRequestStream.filter((pr) -> lastDeliveryDate.before(mapDatePullRequest(pr)));
+        }
+        return findStatus(pullRequestStream.collect(Collectors.toList()));
```
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/web/resources/WarningResource.java]: Maybe extract the retrieval of recent pull requests to a separate method `retrieveRecentPullRequests(group)`.

File: src/main/java/nl/tudelft/ewi/devhub/server/web/resources/WarningResource.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    public static PullRequestWarningResponse findStatus(List<PullRequest> pullRequests){
+        if(pullRequests.stream().filter(PullRequest::isClosed).filter(PullRequest::isMerged).findAny().isPresent()) {
```
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/web/resources/WarningResource.java]: I think only filtering only merged pull requests will do. (They will be closed anyway, maybe we could allow to reopen them later on, but even then, the PR boolean should be the only relevant flag for this check)

File: src/main/java/nl/tudelft/ewi/devhub/server/web/resources/WarningResource.java (hunk 3 above)
Comment (JWGmeligMeyling) [inline src/main/java/nl/tudelft/ewi/devhub/server/web/resources/WarningResource.java]: Even beter, put this logic in the DAO and do the date comparison in SQL.

File: src/test/java/nl/tudelft/ewi/devhub/server/database/controllers/WarningTest.java (hunk 1 above)
Comment (JeremybellEU) [inline src/test/java/nl/tudelft/ewi/devhub/server/database/controllers/WarningTest.java]: :+1: Makes constructing the testing objects a lot easier.

**Comment by JeremybellEU** [comment]
This will conflict with #184 , probably #184 is the improved/second version of this PR?

**Comment by JWGmeligMeyling** [comment]
They are a bit different because I needed HTML in the popover. I'll keep this in mind when merging :wink:

**Comment by JWGmeligMeyling** [comment]
Closing because we have switched to a different approach of generating commit warnings.

**Pull request closed without merging**